nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -436,6 +436,19 @@ bool CollisionMonitor::processStopSlowdown(
         robot_action.req_vel = safe_vel;
         return true;
       }
+    } else {  // Limit
+      Velocity safe_vel;
+      safe_vel.x = std::clamp(velocity.x, -polygon->getLimitX(), polygon->getLimitX());
+      safe_vel.y = std::clamp(velocity.y, -polygon->getLimitY(), polygon->getLimitY());
```

Comment:
For omni-drives the linear velocity could have two OX and OY components simultaneously. There could be an issue when we are limiting each component on a some value, while total maximum allowed speed of robot could exceed this maximum value. For example, the maximum speed in some area is `0.5 m/s`. Omni-drive is moving with `Vx = 0.5` and `Vy = 0.5`. Summary linear speed of robot is `Vlin = ~1.41 * 0.5 = 0.7 m/s` which is higher than limit.

Thus, I'd like recommend to calculate the absolute value of linear velocity, and then restrict each OX/OY component proportionally on clamped value. Something, like made in https://github.com/ros-planning/navigation2/blob/ae887b1eb91bbce5de8f4c8787946acc9af2e384/nav2_dwb_controller/dwb_plugins/src/kinematic_parameters.cpp#L153-L155 but instead of `base_max_vel_x` it will be used current absolute linear speed value.

Reply:
I think it's not the duty of the collision monitor to manage the maximum velocity of the X/Y combination, it's more velocity smoother.

But you are right on one point if we reduce only one component, the motion line will change and that's not good behavior.

I used separate limits to fit the X/Y/TW nav2::velocity design. I am not an expert in omni-drive and I don't know the expected behavior. If the robot is following a wall, it may want to only reduce speed in Y or X axis?

The approach you propose looks good to me. I will implement it.

Reply:
We need to preserve the robot moving direction. If we will reduce only one component of speed, robot will change its direction, which will be at least confusing.
Moreover, usually we are talking about maximum allowed linear and angular velocities for robots in some areas, so that makes also sense from this point of view.